File: app/src/main/java/dev/chu/architecture/java/p87/AppComponent.java
package dev.chu.architecture.java.p87;

import javax.inject.Singleton;

import dagger.BindsInstance;
import dagger.Component;
import dev.chu.architecture.App;

@Component(modules = AppModule.class)
@Singleton
public interface AppComponent {

    MainActivityComponent.Builder mainActivityComponentBuilder();

    void inject(App app);

    @Component.Factory
    interface Factory {
        AppComponent create(@BindsInstance App app, AppModule appModule);
    }

}

File: app/src/main/java/dev/chu/architecture/kotlin/p87/AppComponent.kt
package dev.chu.architecture.kotlin.p87

import dagger.BindsInstance
import dagger.Component
import dev.chu.architecture.App
import javax.inject.Singleton

@Component(modules = [AppModule::class])
@Singleton
interface AppComponent {
    fun mainActivityComponentFactory(): MainActivityComponent.Factory
    fun inject(app: App)

    @Component.Factory
    interface Factory {
        fun create(
            @BindsInstance app: App,
            module: AppModule
        ): AppComponent
    }
}